tensorboard/plugins/text/text_plugin.py

```diff
@@ -203,6 +205,66 @@ def __init__(self, context):
     """
     self._multiplexer = context.multiplexer
 
+    # Cache the last result of index_impl() so that methods that depend on it
+    # can return without blocking (while kicking off a background thread to
+    # recompute the current index).
+    self._index_cached = None
+
+    # Lock that ensures that only one thread attempts to compute index_impl()
+    # at a given time, since it's expensive.
+    self._index_impl_lock = threading.Lock()
+
+    # Pointer to the current thread computing index_impl(), if any.  This is
+    # stored on TextPlugin only to facilitate testing.
+    self._index_impl_thread = None
+
+  def is_active(self):
+    """Determines whether this plugin is active.
+
+    This plugin is only active if TensorBoard sampled any text summaries.
+
+    Returns:
+      Whether this plugin is active.
+    """
+    if not self._multiplexer:
+      return False
+
+    if self._index_cached is not None:
+      # If we already have computed the index, use it to determine whether
+      # the plugin should be active, and if so, return immediately.
+      if any(self._index_cached.values()):
+        return True
+
+    # We haven't conclusively determined if the plugin should be active. Launch
+    # a thread to compute index_impl() and return False to avoid blocking.
+    self._maybe_launch_index_impl_thread()
+    return False
+
+  def _maybe_launch_index_impl_thread(self):
+    """Attempts to launch a thread to compute index_impl().
+
+    This may not launch a new thread if one is already running to compute
+    index_impl(); in that case, this function is a no-op.
+    """
+    # Try to acquire the lock for computing index_impl(), without blocking.
+    if self._index_impl_lock.acquire(False):
+      # We got the lock. Start the thread, which will unlock the lock when done.
+      self._index_impl_thread = threading.Thread(
+          target=self._async_index_impl,
+          name='TextPluginIndexImplThread')
+      self._index_impl_thread.start()
+
+  def _async_index_impl(self):
+    """Computes index_impl() asynchronously on a separate thread."""
+    start = time.time()
+    tf.logging.info('TextPlugin computing index_impl() in a new thread')
+    self._index_cached = self.index_impl()
+    self._index_impl_lock.release()
+    self._index_impl_thread = None
```

Comment:
nit: Should `self._index_impl_thread = None` run before the lock is released? Otherwise, `_index_impl_thread` could be set to a separate thread in between? And we'd be releasing a newly started thread (and leaving the old one open)?

Reply:
Good catch!  You're right, semantically it should definitely be set to None prior to releasing the lock. That said, _index_impl_thread actually only exists to facilitate testing (it's never read outside the test code) so in practice it's a wash.  But fixed anyway.

Reply:
Oh yeah - true. That would've been innocuous.